### ISSUE TITLE
Typechecking lightgbm regression since 1.1.325

### ISSUE THREAD

**Issue by rymndhng** [microsoft/pyright#6224]
**Describe the bug**

Typechecking code from `lightgbm` is slow on 1.1.332 compared to 1.1.325.

**Code or Screenshots**

1. Create a venv and install lightgbm:

```
pip install lightgbm==4.1.0
```

2. Typecheck this code:

```python
import lightgbm as lgb

target = lgb.Dataset()
target.pandas_categorical
```

On 1.1.325, typechecking takes ~2 seconds: 

```
PYRIGHT_PYTHON_FORCE_VERSION=1.1.325 pyright --verbose repro.py
```

On 1.1.332, typechecking about ~5 minutes: 

```
PYRIGHT_PYTHON_FORCE_VERSION=1.1.332 pyright --verbose repro.py
```

Here are the verbose logs from 1.1.322:

```
Could not resolve source for '._ufunc' in file '/Users/raymond.huang/projects/pyright-repro/my_env/lib/python3.11/site-packages/numpy/_typing/__init__.py'
  Attempting to resolve relative import
  Attempting to resolve using root path '/Users/raymond.huang/projects/pyright-repro/my_env/lib/python3.11/site-packages/numpy/_typing'
  Resolved import with file '/Users/raymond.huang/projects/pyright-repro/my_env/lib/python3.11/site-packages/numpy/_typing/_ufunc.pyi'
  Attempting to resolve using root path '/Users/raymond.huang/projects/pyright-repro/my_env/lib/python3.11/site-packages/numpy/_typing'
[FG] Long operation: checking: /Users/raymond.huang/projects/pyright-repro/repro.py (285333ms)
[FG] Long operation: analyzing: /Users/raymond.huang/projects/pyright-repro/repro.py (285501ms)
pyright 1.1.332
/Users/raymond.huang/projects/pyright-repro/repro.py
  /Users/raymond.huang/projects/pyright-repro/repro.py:2:10 - error: Argument missing for parameter "data" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 285.858sec
```

**VS Code extension or command-line**

Command line

**Comment by erictraut** [comment]
I'm unfortunately not able to install lightgbm on my machine. It apparently installs only on Windows and Linux machines, and I have a Mac. Without a repro, it will be difficult to diagnose this.

A lot changed between pyright 1.1.325 and 1.1.332. Are you able to narrow it down to a particular version?

**Comment by rymndhng** [comment]
It appears to have regressed twice:

It was first introduced in 1.1.326, with an execution time of ~60 seconds.
Then, in 1.1.328 this increased again to ~5 minutes.

> I'm unfortunately not able to install lightgbm on my machine. It apparently installs only on Windows and Linux machines, and I have a Mac. Without a repro, it will be difficult to diagnose this.

I'm testing this on a venv on my mac (tested on both m1 and intel macs) with these instructions:

```
python3 -m venv myenv
source myenv/bin/activate
pip install lightgbm==4.1.0
pip install pyright
```

**Comment by erictraut** [comment]
> I'm testing this on a venv on my mac

I suspect it won't build on my machine because I'm on an M2-based (Apple Silicon) Mac. Looks like the library isn't compatible with this. I'm guessing that you're on an x86-base Mac? Here's the error I see when attempting to install the library:
```
      *** Building project with Ninja...
      ninja: error: '/opt/homebrew/opt/libomp/lib/libomp.dylib', needed by '/private/var/folders/nm/2vffx3255wb93j5y5k6yjbyh0000gn/T/pip-install-8k4qwf2w/lightgbm_c7948f6d395e4130a5d3d576a8987cc1/lib_lightgbm.so', missing and no known rule to make it
      
      *** CMake build failed
```

I've reviewed all the changes between pyright 1.1.327 and 1.1.328, and I don't see any that could explain a large perf slowdown. I updated the typeshed stubs in that release, so it's possible that it's a side effect of a change in the stubs. Those are rare.

There were a couple of changes between 1.1.325 and 1.1.326 that are potential candidates but no smoking guns. I also updated the typeshed stubs in 1.1.326.

Without a way to repro this locally, it's going to be difficult to address this.

@debonte, @rchiodo, do either of you have a machine where this library installs without error? If so, are you able to repro the issue as reported above?

**Comment by rchiodo** [comment]
Yeah, this is reproducible for me on a windows machine.

1.1325 takes about 6 seconds
1.1332 is still running after 5 minutes

**Comment by rchiodo** [comment]
For me 1.1.326 is where everything gets slow. It basically never returns for me. For 1.1.325 it comes back in about 5 seconds.

**Comment by erictraut** [comment]
@rchiodo, thanks for confirming. After wrestling with the library a bit, I was able to get it installed, and I'm now able to repro. I have it narrowed down to a particular commit. I'll investigate further.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by rymndhng** [comment]
Fantastic, thanks @erictraut!

**Comment by erictraut** [comment]
We'll keep the issue open until the fix is released.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.334, which I just published. It will also be included in a future release of pylance.
